File: frontend/src/api/task.ts
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import apiClient from './client'

// 定时任务相关接口
export interface ScheduledTask {
  id: number
  name: string
  description: string
  cron_expression: string
  template_id?: number
  template_name?: string
  content: string
  type: string
  asset_ids: number[]
  timeout: number
  enabled: boolean
  update_assets: boolean // 是否更新资产信息
  last_run_at?: string
  next_run_at?: string
  last_status?: string
  created_by: number
  creator_name?: string
  created_at: string
  updated_at: string
}

export interface CreateScheduledTaskRequest {
  name: string
  description?: string
  cron_expression: string
  template_id?: number
  content: string
  type?: string
  asset_ids: number[]
  timeout?: number
  enabled?: boolean
  update_assets?: boolean // 是否更新资产信息
}

export interface UpdateScheduledTaskRequest {
  name?: string
  description?: string
  cron_expression?: string
  template_id?: number
  content?: string
  type?: string
  asset_ids?: number[]
  timeout?: number
  enabled?: boolean
  update_assets?: boolean // 是否更新资产信息
}

export interface ListScheduledTasksResponse {
  data: ScheduledTask[]
  total: number
  page: number
  size: number
}

export interface ScheduledTaskExecution {
  id: number
  scheduled_task_id: number
  asset_id: number
  asset?: {
    id: number
    hostname: string
    ip: string
  }
  trigger_type: string
  status: string
  output?: string
  error?: string
  exit_code?: number
  started_at?: string
  finished_at?: string
  created_at: string
}

export interface ValidateCronResponse {
  valid: boolean
  next_run_at?: string
  error?: string
}

// 定时任务 API
export const scheduledTaskApi = {
  // 获取列表
  list: (page?: number, pageSize?: number, enabled?: boolean) => {
    const params = new URLSearchParams()
    if (page) params.append('page', String(page))
    if (pageSize) params.append('page_size', String(pageSize))
    if (enabled !== undefined) params.append('enabled', String(enabled))
    const query = params.toString()
    return apiClient.get<ListScheduledTasksResponse>(`/tasks${query ? `?${query}` : ''}`)
  },

  // 获取详情
  get: (id: number) => apiClient.get<ScheduledTask>(`/tasks/${id}`),

  // 创建
  create: (data: CreateScheduledTaskRequest) =>
    apiClient.post<ScheduledTask>('/tasks', data),

  // 更新
  update: (id: number, data: UpdateScheduledTaskRequest) =>
    apiClient.put<ScheduledTask>(`/tasks/${id}`, data),

  // 删除
  delete: (id: number) => apiClient.delete(`/tasks/${id}`),

  // 启用
  enable: (id: number) => apiClient.post(`/tasks/${id}/enable`),

  // 禁用
  disable: (id: number) => apiClient.post(`/tasks/${id}/disable`),

  // 获取执行历史
  getExecutions: (id: number, page?: number, pageSize?: number) => {
    const params = new URLSearchParams()
    if (page) params.append('page', String(page))
    if (pageSize) params.append('page_size', String(pageSize))
    const query = params.toString()
    return apiClient.get<{ data: ScheduledTaskExecution[]; total: number }>(
      `/tasks/${id}/executions${query ? `?${query}` : ''}`
    )
  },

  // 验证 Cron 表达式
  validateCron: (cronExpression: string) =>
    apiClient.get<ValidateCronResponse>(
      `/tasks/validate-cron?cron_expression=${encodeURIComponent(cronExpression)}`
    ),

  // 导出配置
  exportConfig: () => apiClient.get<ExportScheduledTasksConfig>('/tasks/export'),

  // 导入配置
  importConfig: (data: ImportScheduledTasksConfig) =>
    apiClient.post<ImportScheduledTasksResult>('/tasks/import', data),
}

// 导入导出相关接口
export interface ExportScheduledTaskConfig {
  name: string
  description: string
  cron_expression: string
  template_name?: string
  content: string
  type: string
  timeout: number
  enabled: boolean
  update_assets: boolean
  target_hosts: string[]
}

export interface ExportScheduledTasksConfig {
  version: string
  export_at: string
  tasks: ExportScheduledTaskConfig[]
}

export interface ImportScheduledTaskConfig {
  name: string
  description?: string
  cron_expression: string
  template_name?: string
  content?: string
  type?: string
  timeout?: number
  enabled?: boolean
  update_assets?: boolean
  target_hosts?: string[]
}

export interface ImportScheduledTasksConfig {
  version: string
  tasks: ImportScheduledTaskConfig[]
}

export interface ImportScheduledTasksResult {
  total: number
  success: number
  failed: number
  errors?: string[]
  skipped?: string[]
}

// 导出默认 API
export default scheduledTaskApi

File: backend/internal/model/task.go
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package model

import (
	"time"

	"gorm.io/gorm"
)

// TaskTemplate represents a task template
type TaskTemplate struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	Name        string         `gorm:"not null;size:100" json:"name"`
	Description string         `gorm:"type:text" json:"description"`
	Content     string         `gorm:"type:text" json:"content"` // Script or command content
	Type        string         `gorm:"size:50" json:"type"`      // shell, python, etc.
	CreatedBy   uint           `json:"created_by"`
	Creator     User           `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	// Relationships
	Tasks []Task `gorm:"foreignKey:TemplateID" json:"tasks,omitempty"`
}

// Task represents an execution task
type Task struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	TemplateID  *uint          `json:"template_id"`
	Template    *TaskTemplate  `gorm:"foreignKey:TemplateID" json:"template,omitempty"`
	Name        string         `gorm:"not null;size:100" json:"name"`
	Description string         `gorm:"type:text" json:"description"`
	Content     string         `gorm:"type:text" json:"content"`                    // Script or command content
	Type        string         `gorm:"size:50" json:"type"`                         // shell, python, etc.
	Timeout     int            `gorm:"default:600" json:"timeout"`                  // Execution timeout in seconds (default 10 minutes)
	Status      string         `gorm:"default:pending;size:20;index" json:"status"` // pending, running, success, failed, cancelled
	AssetIDs    string         `gorm:"type:text" json:"asset_ids_str"`              // Comma-separated asset IDs for execution
	CreatedBy   uint           `json:"created_by"`
	Creator     User           `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	StartedAt   *time.Time     `json:"started_at"`
	FinishedAt  *time.Time     `json:"finished_at"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	// Relationships
	Executions []TaskExecution `gorm:"foreignKey:TaskID" json:"executions,omitempty"`
}

// TaskExecution represents a task execution on a specific host
type TaskExecution struct {
	ID              uint           `gorm:"primaryKey" json:"id"`
	TaskID          *uint          `gorm:"index" json:"task_id,omitempty"`                      // 关联的执行任务 ID（可为空）
	Task            *Task          `gorm:"foreignKey:TaskID" json:"task,omitempty"`
	ScheduledTaskID *uint          `gorm:"index" json:"scheduled_task_id,omitempty"`            // 关联的定时任务 ID（可为空）
	AssetID         uint           `gorm:"not null;index" json:"asset_id"`
	Asset           Asset          `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
	TriggerType     string         `gorm:"size:20;default:manual" json:"trigger_type"`          // manual, scheduled
	Status          string         `gorm:"default:pending;size:20;index" json:"status"`         // pending, running, success, failed, cancelled
	ExitCode        *int           `json:"exit_code"`
	Output          string         `gorm:"type:text" json:"output"`                             // Command output
	Error           string         `gorm:"type:text" json:"error"`                              // Error message
	StartedAt       *time.Time     `json:"started_at"`
	FinishedAt      *time.Time     `json:"finished_at"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
}
